### PR TITLE
refactor: deprecate soft opt out

### PR DESCRIPTION
**Pull request by insumity**
## Description

Closes: #XXXX

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] Added `!` to the type prefix if the change is [state-machine breaking](https://github.com/cosmos/interchain-security/blob/main/RELEASES.md#breaking-changes)
* [ ] Confirmed this PR does not introduce changes requiring state migrations, OR migration code has been added to consumer and/or provider modules
* [ ] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
* [ ] Provided a link to the relevant issue or specification
* [ ] Followed the guidelines for [building SDK modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/00-intro.md)
* [ ] Included the necessary unit and integration [tests](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#testing)
* [ ] Added a changelog entry to `CHANGELOG.md`
* [ ] Included comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] Updated the relevant documentation or specification
* [ ] Reviewed "Files changed" and left comments if necessary 
* [ ] Confirmed all CI checks have passed
* [ ] If this PR is library API breaking, bump the go.mod version string of the repo, and follow through on a new major release

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` the type prefix if the change is state-machine breaking
* [ ] confirmed this PR does not introduce changes requiring state migrations, OR confirmed migration code has been added to consumer and/or provider modules
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage

### CODE REVIEW

File: tests/e2e/steps_partial_set_security.go [hunk 1]
```diff
@@ -1506,8 +1482,10 @@ func stepsValidatorsAllowlistedChain() []Step {
 				InitialHeight: clienttypes.Height{RevisionNumber: 0, RevisionHeight: 1},
 				TopN:          0,
 				// only "alice" and "bob" are allowlisted (see `getDefaultValidators` in `tests/e2e/config.go`)
-				Allowlist: []string{"cosmosvalcons1qmq08eruchr5sf5s3rwz7djpr5a25f7xw4mceq",
-					"cosmosvalcons1nx7n5uh0ztxsynn4sje6eyq2ud6rc6klc96w39"},
+				Allowlist: []string{
```
Comment (insumity) [inline tests/e2e/steps_partial_set_security.go:1485]: Those and similar changes stem from running `make format`.

**Review by p-offtermatt — APPROVED** [review]
LGTM.

Wondering if deprecating like this is actually easier than having a migration?

File: .changelog/unreleased/improvements/consumer/1964-deprecate-soft-opt-out.md [hunk 2]
```diff
@@ -0,0 +1,2 @@
+- Deprecate soft opt-out because after the release of PSS it is not needed
+  anymore. ([\#1964](https://github.com/cosmos/interchain-security/pull/1964))
```
Comment (mpoke) [inline .changelog/unreleased/improvements/consumer/1964-deprecate-soft-opt-out.md:2]:
```suggestion
- Remove soft opt-out feature. 
  ([\#1964](https://github.com/cosmos/interchain-security/pull/1964))
```

File: docs/docs/adrs/adr-009-soft-opt-out.md [hunk 3]
```diff
@@ -10,8 +10,8 @@ title: Soft Opt-Out
 
 ## Status
 
-Accepted
-
+Deprecated
+Deprecated by [Partial Set Security](adr-015-partial-set-security.md)
```
Comment (mpoke) [inline docs/docs/adrs/adr-009-soft-opt-out.md:14]:
```suggestion
Deprecated (by [Partial Set Security](adr-015-partial-set-security.md))
```

Comment (mpoke) [reply docs/docs/adrs/adr-009-soft-opt-out.md:14]: Also, add an entry to the above changelog, i.e., 
```
* 6/18/24: Change status to deprecated
```

File: x/ccv/consumer/types/keys.go [hunk 4]
```diff
@@ -65,8 +65,9 @@ const (
 	// [DEPRECATED]
 	LastStandaloneHeightByteKey
 
-	// SmallestNonOptOutPowerByteKey is the byte that will store the smallest val power that cannot opt out
-	SmallestNonOptOutPowerByteKey
+	// NOTE: This key is deprecated, but left in place to avoid consumer state migrations
+	// [DEPRECATED]
+	DeprecatedSmallestNonOptOutPowerByteKey
```
Comment (mpoke) [inline x/ccv/consumer/types/keys.go:70]: Cleanup the state in a migration.

**Review by mpoke — COMMENTED** [review]
Great work @insumity. See my comments below.

File: x/ccv/consumer/types/keys.go (hunk 4 above)
Comment (MSalopek) [reply x/ccv/consumer/types/keys.go:70]: The state is 2 numbers (neutron and stride)
The migration can be skipped so we don't have to bump consensus versions .

Comment (mpoke) [reply x/ccv/consumer/types/keys.go:70]: I think we already have to bump consensus versions for ICS 4.3. If we do, then let's just cleanup the state. Otherwise, just ignore.

File: tests/e2e/steps_consumer_misbehaviour.go [hunk 5]
```diff
@@ -173,7 +172,7 @@ func stepsStartChainsWithSoftOptOut(consumerName string) []Step {
 				ChainID(consumerName): ChainState{
 					ValPowers: &map[ValidatorID]uint{
 						ValidatorID("alice"): 500,
-						ValidatorID("bob"):   20,
+						ValidatorID("bob"):   0,
```
Comment (sainoe) [inline tests/e2e/steps_consumer_misbehaviour.go:175]: Is it expected? Bob shouldn't be in the consunmer's valset even with 0 vp, right?

File: tests/e2e/steps_consumer_misbehaviour.go [hunk 6]
```diff
@@ -189,7 +188,7 @@ func stepsStartChainsWithSoftOptOut(consumerName string) []Step {
 				ChainID(consumerName): ChainState{
 					ValPowers: &map[ValidatorID]uint{
 						ValidatorID("alice"): 511,
-						ValidatorID("bob"):   20,
+						ValidatorID("bob"):   0,
```
Comment (sainoe) [inline tests/e2e/steps_consumer_misbehaviour.go:191]: ditto

File: tests/e2e/steps_consumer_misbehaviour.go [hunk 7]
```diff
@@ -113,12 +118,6 @@ func stepsStartChainsWithSoftOptOut(consumerName string) []Step {
 				Validators: []StartChainValidator{
 					{Id: ValidatorID("alice"), Stake: 500000000, Allocation: 10000000000},
 				},
-				// For consumers that're launching with the provider being on an earlier version
-				// of ICS before the soft opt-out threshold was introduced, we need to set the
-				// soft opt-out threshold to 0.05 in the consumer genesis to ensure that the
-				// consumer binary doesn't panic. Sdk requires that all params are set to valid
-				// values from the genesis file.
-				GenesisChanges: ".app_state.ccvconsumer.params.soft_opt_out_threshold = \"0.05\"",
 			},
```
Comment (sainoe) [inline tests/e2e/steps_consumer_misbehaviour.go:121]: Could you update the comment above on the lines 113-114?

File: x/ccv/consumer/types/keys.go (hunk 4 above)
Comment (insumity) [reply x/ccv/consumer/types/keys.go:70]: We already bump the consensus version of the `provider` module but **not** of the `consumer` module. Talked with @MSalopek about this and it would make things substantially more complicated in regards to ICS 5.0 because ICS 5.0 has already bumped the consumer to version 3.

File: x/ccv/types/params.go [hunk 8]
```diff
@@ -83,10 +79,11 @@ func NewParams(enabled bool, blocksPerDistributionTransmission int64,
 		ConsumerRedistributionFraction:    consumerRedistributionFraction,
 		HistoricalEntries:                 historicalEntries,
 		UnbondingPeriod:                   consumerUnbondingPeriod,
-		SoftOptOutThreshold:               softOptOutThreshold,
-		RewardDenoms:                      rewardDenoms,
-		ProviderRewardDenoms:              providerRewardDenoms,
-		RetryDelayPeriod:                  retryDelayPeriod,
+		// DEPRECATED but setting here to 0 (i.e., disabled) for older versions of interchain-security
```
[the remark this answers is not in the document]
Comment (mpoke) [reply x/ccv/types/params.go:82]: What's the difference between deprecating a field and reserving it? cc @MSalopek

[the remark this answers is not in the document]
Comment (MSalopek) [reply x/ccv/types/params.go:82]: I think you're mixing apples and oranges.

This section is about setting something to 0 in go code. "reserved" is a term from protobuf.

"reserved" keyword in protobuf is documente here:
* https://protobuf.dev/programming-guides/proto3/#reserved
It helps you prevent somebody from accidentally using an enum that was deprecated in an earlier version. It also helps you avoid issues that arise when you simply remove a field and when you accidentally re-map enums in a breaking manner.

File: proto/interchain_security/ccv/v1/shared_consumer.proto [hunk 9]
```diff
@@ -62,11 +62,8 @@ message ConsumerParams {
     google.protobuf.Duration unbonding_period = 9
         [ (gogoproto.nullable) = false, (gogoproto.stdduration) = true ];
 
-    // The threshold for the percentage of validators at the bottom of the set who
-    // can opt out of running the consumer chain without being punished. For
-    // example, a value of 0.05 means that the validators in the bottom 5% of the
-    // set can opt out
-    string soft_opt_out_threshold = 10;
+    // !!! DEPRECATED !!! soft_opt_out_threshold is deprecated. see docs/docs/adrs/adr-015-partial-set-security.md
+    string soft_opt_out_threshold = 10 [deprecated = true];
```
Comment (MSalopek) [inline proto/interchain_security/ccv/v1/shared_consumer.proto:66]: `ConsumerParams` is used to send data to the consumer chain "over the wire" - it is used in the consumer genesis.

Setting it to deprecated allows the field to exist but it will always be empty, thus maintaining backward compatibility.

We will not need to write a special parser to go from v4.2.0 `ConsumerParams` to some future version `ConsumerParams`.

File: tests/e2e/actions.go [hunk 10]
```diff
@@ -606,6 +605,11 @@ func (tr *TestConfig) startConsumerChain(
 	verbose bool,
 ) {
 	fmt.Println("Starting consumer chain ", action.ConsumerChain)
+
+	// Note that Soft Opt-Out has been deprecated. Nevertheless, the parameter still exists and is expected to be set
+	// because the SDK requires that all params are set to valid values in the genesis file.
+	action.GenesisChanges = action.GenesisChanges + ".app_state.ccvconsumer.params.soft_opt_out_threshold = \"0\""
```
Comment (MSalopek) [inline tests/e2e/actions.go:611]: Why do you need to set it? You can delete this line.

File: tests/e2e/actions.go [hunk 11]
```diff
@@ -838,6 +842,10 @@ func (tr TestConfig) changeoverChain(
 		log.Fatal(err, "\n", string(bz))
 	}
 
+	// Note that Soft Opt-Out has been deprecated. Nevertheless, the parameter still exists and is expected to be set
+	// because the SDK requires that all params are set to valid values in the genesis file.
+	action.GenesisChanges = action.GenesisChanges + ".app_state.ccvconsumer.params.soft_opt_out_threshold = \"0\""
```
Comment (MSalopek) [inline tests/e2e/actions.go:847]: Same as above

File: tests/e2e/step_delegation.go [hunk 12]
```diff
@@ -368,9 +263,9 @@ func stepsRedelegateShort(consumerName string) []Step {
 				ChainID(consumerName): ChainState{
 					ValPowers: &map[ValidatorID]uint{
 						// Now power changes are seen by consumer
-						ValidatorID("alice"): 509,
+						ValidatorID("alice"): 910,
```
Comment (MSalopek) [inline tests/e2e/step_delegation.go:266]: Why has the ratio changed so much?

Are you trying to check that the soft opt-out no longer works?

File: tests/e2e/steps_active_set_changes.go [hunk 13]
```diff
@@ -86,12 +86,6 @@ func stepsActiveSetChanges() []Step {
 					{Id: ValidatorID("bob"), Stake: 200000000, Allocation: 10000000000},
 					{Id: ValidatorID("carol"), Stake: 700000000, Allocation: 10000000000},
 				},
-				// For consumers that're launching with the provider being on an earlier version
```
Comment (MSalopek) [inline tests/e2e/steps_active_set_changes.go:89]: This test removes the threshold while the ones above have it stil

File: x/ccv/types/params.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (mpoke) [reply x/ccv/types/params.go:82]: So, is it better to remove the field and mark it as reserved or use the deprecated annotation?

File: tests/e2e/actions.go (hunk 10 above)
Comment (MSalopek) [reply tests/e2e/actions.go:611]: The default value is `""` which is a valid zero string.

File: tests/e2e/actions.go [hunk 14]
```diff
@@ -2065,7 +2073,7 @@ func (tr TestConfig) invokeDoublesignSlash(
 		if err != nil {
 			log.Fatal(err, "\n", string(bz))
 		}
-		tr.waitBlocks("provi", 10, 2*time.Minute)
+		tr.waitBlocks("provi", 20, 4*time.Minute)
```
Comment (insumity) [inline tests/e2e/actions.go:2076]: Just a note for posterity: Increased this due to some flakiness in the test involving double signing.

File: tests/e2e/step_delegation.go (hunk 12 above)
Comment (insumity) [reply tests/e2e/step_delegation.go:266]: > Why has the ratio changed so much?

TL;DR because we removed opt-out related steps that were reducing the `ValPower` of `alice`.

Specifically, we removed the `stepsRedelegateForOptOut` and `stepsDowntimeWithOptOut`. The `stepsRedelegateForOptOut` was the one that [reduced](https://github.com/cosmos/interchain-security/blob/060c943b2f17b2696e668c669352a366d3610d5d/tests/e2e/step_delegation.go#L223) the `ValPower` of `alice` to `60` so it could test the opt out functionality.

Before, in the `happyPathSteps` we had:
```
...
stepsRedelegateForOptOut("consu"),
stepsDowntimeWithOptOut("consu"),
stepsRedelegate("consu"),
...
```
but we removed `stepsRedelegateForOptOut` and `stepsDowntimeWithOptOut` and hence now when `stepsRedelegate` is called, the `ValPower` of `alice` is not `60` anymore but `510`. Redelegating `400` to `alice` makes `alice` have a `ValPower` `910`.

> Are you trying to check that the soft opt-out no longer works?
No.

File: tests/e2e/actions.go (hunk 10 above)
Comment (insumity) [reply tests/e2e/actions.go:611]: > Why do you need to set it? You can delete this line.

I cannot delete this line. If I do, I get the following error:
```
2024/06/21 11:02:14 getBalance() failed: /usr/local/bin/docker exec interchain-security-instance interchain-security-cd query bank balances consumer1sx6j9g2rh324a342a5f0rnx7me34r9nwduz5te --node tcp://7.7.8.253:26658 -o json: exit status 1
Error: post failed: Post "http://7.7.8.253:26658": dial tcp 7.7.8.253:26658: connect: no route to host
```
because the consumer chain never starts. Although we do not utilize `soft_opt_threshold`, the parameter has still to be set. This is why this line existed in all other E2E tests as well. In this PR, I just moved this line to a central place so we do not have to repeat it every time we create a new consumer chain.

> The default value is "" which is a valid zero string.

Yes, we could potentially set simply to `""` here but setting to `0` makes it clear that the soft opt-out feature is disabled.

File: tests/e2e/actions.go (hunk 11 above)
Comment (insumity) [reply tests/e2e/actions.go:847]: See my [response](https://github.com/cosmos/interchain-security/pull/1964/files/#r1648659182).

File: tests/e2e/steps_active_set_changes.go (hunk 13 above)
Comment (insumity) [reply tests/e2e/steps_active_set_changes.go:89]: Threshold are deleted from individual `StarConsumerChainAction` and are moved to `startConsumerChain` so we do not have to set the threshold everywhere. This was just part of small refactoring.

File: tests/e2e/step_delegation.go (hunk 12 above)
Comment (MSalopek) [reply tests/e2e/step_delegation.go:266]: Thank you for elaborating. It was not clear to me after going over it a couple times.

File: proto/interchain_security/ccv/v1/shared_consumer.proto (hunk 9 above)
Comment (insumity) [reply proto/interchain_security/ccv/v1/shared_consumer.proto:66]: Agreed. This is why we did not use `reserved`.

File: x/ccv/types/params.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (MSalopek) [reply x/ccv/types/params.go:82]: See discussion: https://github.com/cosmos/interchain-security/pull/1964#discussion_r1647272147

File: tests/e2e/steps_start_chains.go [hunk 15]
```diff
@@ -159,12 +159,6 @@ func stepsStartConsumerChain(consumerName string, proposalIndex, chainIndex uint
 					{Id: ValidatorID("alice"), Stake: 500000000, Allocation: 10000000000},
 					{Id: ValidatorID("carol"), Stake: 500000000, Allocation: 10000000000},
 				},
-				// For consumers that're launching with the provider being on an earlier version
```
Comment (MSalopek) [inline tests/e2e/steps_start_chains.go:162]: thanks for figuring out how to remove this.

Much appreciated!

**Review by MSalopek — APPROVED** [review]
Thanks for spending the time on this.

I'm glad you managed to clean up the state and simplify the protocol.

Approve!

File: tests/e2e/steps_start_chains.go (hunk 15 above)
Comment (insumity) [reply tests/e2e/steps_start_chains.go:162]: Thanks for pointing it out 😃
